### PR TITLE
Feature/integrate vedtakoppsummering dokgen

### PR DESCRIPTION
**Pull request by MengZhuNAV**
Integrate preview with docgen

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -168,6 +168,11 @@
 			<groupId>org.springframework.retry</groupId>
 			<artifactId>spring-retry</artifactId>
 		</dependency>
+		<dependency>
```
Comment (henninghaakonsen) [inline pom.xml]: I dont think we need this - probably from the proxy rest template Tor added.

File: src/main/kotlin/no/nav/familie/ba/sak/behandling/FagsakController.kt [hunk 2]
```diff
@@ -51,14 +53,18 @@ class FagsakController (
...
         val saksbehandlerId = oidcUtil.getClaim("preferred_username")
-        FagsakController.logger.info("{} henter vedtaksbrev", saksbehandlerId ?: "VL")
+        logger.info("{} henter vedtaksbrev", saksbehandlerId ?: "VL")
 
-        //TODO: integration with DokGen
+        val behandlingVedtak = fagsakService.hentVedtakForBehandling(behandlingId)
+                ?: return Ressurs.failure("Vedtak ikke funnet")
 
-        return Ressurs.success("<H1>Vedtaksbrev</H1><br/>FagsakID= $fagsakId<br /><P>Backend API not implemented yet</P>");
+        val html = docgenKlient.lagHtmlFraMarkdown(behandlingVedtak.stønadBrevMarkdown)
```
Comment (henninghaakonsen) [inline src/main/kotlin/no/nav/familie/ba/sak/behandling/FagsakController.kt]: What happens if this api call failes? Maybe let the service return Ressurs, or wrap the service call with `Result.runCatching`.

File: src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+@Service
+class DokGenKlient @Autowired constructor(
+    @Value("\${FAMILIE_BA_DOKGEN_API_URL}") private val dokgenServiceUri: String,
+    private val restTemplateMedProxy: RestTemplate) {
+
+    fun lagHtmlFraMarkdown(markdown: String): String {
+        val url = URI.create(dokgenServiceUri + "/template/Innvilget/preview-html/Innvilget1") // Kan laste opp og erstatte "Innvilget" med "Fritekstmal" f.eks
+        val response = utførRequest(POST, String::class.java, url, markdown)
+
+        if (!response.statusCode.is2xxSuccessful) {
```
Comment (stigebil) [inline src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt]: This check is unnecessary. Default behaviour for restTemplate is to throw a RestClientException if status code is different than 2xx.

File: src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            throw RuntimeException(response.toString()) //TODO feilhåndtering
+        }
+        return response.body.orEmpty()
+    }
+
+    fun hentMarkdownForMal(malNavn: String) : String {
+        val url = URI.create(dokgenServiceUri + "/template/" + malNavn + "/markdown")
+        val response = utførRequest(GET, String::class.java, url)
+
+        if (!response.statusCode.is2xxSuccessful) {
```
Comment (stigebil) [inline src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt]: This check is also unnecessary.

File: src/main/kotlin/no/nav/familie/ba/sak/behandling/FagsakService.kt [hunk 5]
```diff
@@ -50,4 +53,8 @@ class FagsakService(
     fun lagreFagsak(fagsak: Fagsak) {
         fagsakRepository.save(fagsak)
     }
+
+    fun hentVedtakForBehandling(behandlingId: Long): BehandlingVedtak?{
+        return behandlingVedtakRepository.finnBehandlingVedtak(behandlingId)
```
Comment (henninghaakonsen) [inline src/main/kotlin/no/nav/familie/ba/sak/behandling/FagsakService.kt]: I think this fits better inside `BehandlingslagerService`.

File: src/main/kotlin/no/nav/familie/ba/sak/config/NaisProxyCustomizer.kt [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+import org.apache.http.HttpRequest
+import org.apache.http.client.HttpClient
+import org.apache.http.impl.client.HttpClientBuilder
+import org.apache.http.impl.conn.DefaultProxyRoutePlanner
+import org.apache.http.protocol.HttpContext
+import org.springframework.boot.web.client.RestTemplateCustomizer
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory
+import org.springframework.web.client.RestTemplate
+
+class NaisProxyCustomizer : RestTemplateCustomizer {
```
Comment (henninghaakonsen) [inline src/main/kotlin/no/nav/familie/ba/sak/config/NaisProxyCustomizer.kt]: We dont need this.

File: src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+class DokGenKlient @Autowired constructor(
+    @Value("\${FAMILIE_BA_DOKGEN_API_URL}") private val dokgenServiceUri: String,
+    private val restTemplateMedProxy: RestTemplate) {
+
+    fun lagHtmlFraMarkdown(markdown: String): String {
+        val url = URI.create(dokgenServiceUri + "/template/Innvilget/preview-html/Innvilget1") // Kan laste opp og erstatte "Innvilget" med "Fritekstmal" f.eks
+        val response = utførRequest(POST, String::class.java, url, markdown)
+
+        if (!response.statusCode.is2xxSuccessful) {
+            throw RuntimeException(response.toString()) //TODO feilhåndtering
```
Comment (henninghaakonsen) [inline src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt]: I know the code is from Tor's branch, but maybe you can add error handling? The idea should be that the api should always return Ressurs, so throwing an Error will not do the trick I guess. Returning Ressurs here might be a better solution?

File: src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        }
+        return response.body.orEmpty()
+    }
+
+    fun <T> utførRequest(httpMethod: HttpMethod, responseType: Class<T>, requestUrl: URI, requestBody: Any? = null): ResponseEntity<T> {
+        val headers = HttpHeaders()
+        headers.contentType = MediaType.TEXT_PLAIN
```
Comment (stigebil) [inline src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt]: Do this work with Norwegian letters? Or do you need charset like text/plain; charset=UTF-8

File: src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt (hunk 7 above)
Comment (henninghaakonsen) [reply src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt]: @tnarland kanskje dere kan se på dette sammen? Det er ikke ferdig før feilhåndtering er ordentlig på plass.

File: src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt (hunk 8 above)
Comment (henninghaakonsen) [reply src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt]: Nice catch, I do think we need UTF-8 here.

File: src/test/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlientTest.kt [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.web.client.RestTemplate
+
+internal class DokGenKlientTest {
+
+    private val dokgenKlient: DokGenKlient = DokGenKlient(
+        "http://localhost:8080", RestTemplate()
+    )
+
+    @Test
+    fun `Test å hente Markdown og konvertere til html når dokgen kjører lokalt`() {
```
Comment (stigebil) [inline src/test/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlientTest.kt]: possible to write this test without dependency to dokgen? Maybe starting dokgen in Testcontainers? Or using mockingtools like mockwebserver

File: pom.xml [hunk 10]
```diff
@@ -168,6 +168,11 @@
...
+			<groupId>org.apache.httpcomponents</groupId>
+			<artifactId>httpclient</artifactId>
```
Comment (stigebil) [inline pom.xml]: Where is this dependency used?

**Comment by henninghaakonsen** [comment]
Please also check your test-coverage. It is now at 14% for new code. I am not very good at this my self, but try to bump it a bit :) @MengZhuNAV

File: src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt (hunk 8 above)
Comment (tnarland) [reply src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt]: We have a RestTemplateConfig with an added
StringHttpMessageConverter(StandardCharsets.UTF_8)

File: pom.xml (hunk 1 above)
Comment (tnarland) [reply pom.xml]: Yes, I forgot to remove it 🤦‍♂

File: src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt (hunk 7 above)
Comment (tnarland) [reply src/main/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlient.kt]: Enig i det.. kan se på det, men dette er teknisk gjeld som passer å se på etter demoen, tenker jeg, med mindre vi får tid til overs

File: src/test/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlientTest.kt (hunk 9 above)
Comment (tnarland) [reply src/test/kotlin/no/nav/familie/ba/sak/vedtak/DokGenKlientTest.kt]: will look into that 👍

File: pom.xml (hunk 10 above)
Comment (tnarland) [reply pom.xml]: nowhere 👍

**Pull request merged**